### PR TITLE
Add format validation

### PR DESCRIPTION
**Pull request by polomani**
- test for special characters
- test if qudt URL is valid

my first pr in kotlin :)

### CODE REVIEW

File: src/test/resources/duplicateExternalId.json [hunk 1]
```diff
@@ -52,7 +52,7 @@
       "offset": 273.15
     },
     "source": "qudt.org",
-    "sourceReference": "http://qudt.org/vocab/unit/DEG_C",
+    "sourceReference": "http://qudt.org/vocab/unit/DEG",
```
Comment (matiascognite) [inline src/test/resources/duplicateExternalId.json:55]: Technically, the sourceReference in line 161 should be changed as well, but the verification crashes before it is detected, so it's ok

Comment (polomani) [reply src/test/resources/duplicateExternalId.json:55]: agree

File: src/main/kotlin/com/cognite/units/UnitService.kt [hunk 2]
```diff
@@ -69,6 +96,11 @@ class UnitService(unitsPath: URL, systemPath: URL) {
                 "Invalid externalId ${it.externalId} for unit ${it.name} (${it.quantity})"
             }
 
+            // if source is qudt.org, reference should be in the format http://qudt.org/vocab/unit/{unit.name}
+            assert(it.sourceReference?.equals(generatedExpectedSourceReference(it)) ?: true) {
```
Comment (matiascognite) [inline src/main/kotlin/com/cognite/units/UnitService.kt:100]: do we want the reverse? If the reference contains qudt.org, then the source should be qudt.org. For some reason, this is not true for Kelvin

Could have something along the lines of:
```
            val sourceContainsQudt = it.source.equals("qudt.org")
            val sourceReferenceContainsQudt = it.sourceReference?.contains("qudt.org") ?: false
            assert(sourceContainsQudt == sourceReferenceContainsQudt) {
                "Qudt: Inconsistent source ${it.source} and sourceReference ${it.sourceReference} for unit ${it.name}"
            }
```

File: src/main/kotlin/com/cognite/units/UnitService.kt [hunk 3]
```diff
@@ -45,12 +46,38 @@ class UnitService(unitsPath: URL, systemPath: URL) {
         loadSystem(systemPath)
     }
 
+    private fun sanitizeIdentifier(identifier: String): String {
+        // remove all special characters except - and _
+        return identifier.toLowerCase().replace(Regex("[^a-z0-9_-]"), "_")
```
Comment (matiascognite) [inline src/main/kotlin/com/cognite/units/UnitService.kt:51]:
```suggestion
        return identifier.lowercase().replace(Regex("[^a-z0-9_-]"), "_")
```

File: src/main/kotlin/com/cognite/units/UnitService.kt [hunk 4]
```diff
@@ -45,12 +46,38 @@ class UnitService(unitsPath: URL, systemPath: URL) {
...
+
+        // check reference is a valid http(s) url if present
+        if (unit.sourceReference != null) {
+            try {
+                val url = URI.create(unit.sourceReference).toURL()
+                if (url.protocol != "http" && url.protocol != "https") {
+                    throw IllegalArgumentException(errorMessage)
+                }
+            } catch (e: Exception) {
+                throw IllegalArgumentException(errorMessage)
```
Comment (matiascognite) [inline src/main/kotlin/com/cognite/units/UnitService.kt:75]:
```suggestion
                throw IllegalArgumentException(errorMessage, e)
```

**Review by matiascognite — APPROVED** [review]
Some nits, otherwise LGTM 👍

File: src/main/kotlin/com/cognite/units/UnitService.kt [hunk 5]
```diff
@@ -69,6 +96,17 @@ class UnitService(unitsPath: URL, systemPath: URL) {
...
+            // if source is qudt.org, reference should be in the format http://qudt.org/vocab/unit/{unit.name}
+            assert(it.sourceReference?.equals(generatedExpectedSourceReference(it)) ?: true) {
+                "Invalid sourceReference ${it.sourceReference} for unit ${it.name} (${it.quantity})"
+            }
+
+            val sourceIsQudt = it.source?.contains("qudt") ?: false
+            val sourceReferenceContainsQudt = it.sourceReference?.contains("qudt.org") ?: false
+            assert(sourceIsQudt == sourceReferenceContainsQudt) {
+                "Qudt: Inconsistent source ${it.source} and sourceReference ${it.sourceReference} for unit ${it.name}"
+            }
```
Comment (polomani) [inline src/main/kotlin/com/cognite/units/UnitService.kt:108]:
```suggestion
            val sourceIsQudt = it.source.equals("qudt.org")
            val sourceReferenceContainsQudt = it.sourceReference?.contains("qudt") ?: false
            assert(sourceIsQudt == sourceReferenceContainsQudt) {
                "Qudt: Inconsistent source ${it.source} and sourceReference ${it.sourceReference} for unit ${it.name}"
            }
```

**Pull request merged**